### PR TITLE
[1 / 3] improvements to saving and loading callback state

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Initial PR for an improved state saving and loading for Callbacks.

1) This is a bugfix, since pickling with the keys `type(callback)` would pickle the callback class source, and this prevents loading the checkpoint **outside** of PL (when PL not installed).

2) Feature: With a state identifier, we will be able to produce a (more) unique key for the callbacks. It allows us to save more than one of the same type of callback since we can now disambiguate. This PR does not yet add unique identifiers. This will come in a follow up part.

This change is fully backward compatible (old checkpoints can still be loaded the same way). 

Part of #6467 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: tests/checkpointing/test_legacy_checkpoints.py [hunk 1]
```diff
@@ -60,6 +60,7 @@
         "1.2.5",
         "1.2.6",
         "1.2.7",
+        "1.2.8",
```
Comment (ananthsub) [inline tests/checkpointing/test_legacy_checkpoints.py:63]: do you know why this is failing the test?

Comment (carmocca) [reply tests/checkpointing/test_legacy_checkpoints.py:63]: Because 1.2.8 hasn't been released yet

File: pytorch_lightning/trainer/callback_hook.py [hunk 2]
```diff
@@ -257,18 +258,22 @@ def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> Dict[Type, dict]:
...
+        version = checkpoint.get('pytorch-lightning_version')
         # Todo: the `callback_states` are dropped with TPUSpawn as they
         # can't be saved using `xm.save`
         # https://github.com/pytorch/xla/issues/2773
         if callback_states is not None:
             for callback in self.callbacks:
-                state = callback_states.get(type(callback))
+                state = (
+                    callback_states.get(callback.state_identifier)
+                    or callback_states.get(callback._legacy_state_identifier)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/callback_hook.py:275]: here is the backward compatibility.

in this PR, I "hardcode" this check here, but I have ideas for a more general way of upgrading checkpoints for backward compatibility. Will discuss in the linked issue.

File: pytorch_lightning/callbacks/base.py [hunk 3]
```diff
@@ -29,6 +29,14 @@ class Callback(abc.ABC):
     Subclass this class and override any of the relevant hooks
     """
 
+    @property
+    def state_identifier(self) -> str:
+        return self.__class__.__name__
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/base.py:34]: from the linked issue, i'm curious what implementation you're going with for the stateful callbacks that come with pytorch lightning. i'm assuming you'll send this out in a separate PR

File: pytorch_lightning/trainer/callback_hook.py [hunk 4]
```diff
@@ -257,18 +257,21 @@ def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> Dict[Type, dict]:
...
+                )
```
Comment (carmocca) [inline pytorch_lightning/trainer/callback_hook.py:274]:
```suggestion
                state = callback_states.get(callback.state_identifier, callback_states.get(callback._legacy_state_identifier))
```

File: pytorch_lightning/trainer/callback_hook.py [hunk 5]
```diff
@@ -257,18 +257,21 @@ def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> Dict[Type, dict]:
...
                 state = callback.on_save_checkpoint(self, self.lightning_module, checkpoint)
             if state:
-                callback_states[type(callback)] = state
+                callback_states[callback.state_identifier] = state
         return callback_states
 
-    def on_load_checkpoint(self, checkpoint):
+    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
         """Called when loading a model checkpoint."""
         callback_states = checkpoint.get('callbacks')
```
Comment (carmocca) [inline pytorch_lightning/trainer/callback_hook.py:265]: Let's annotate this for clarity

```suggestion
        callback_states: Dict[str, dict] = checkpoint.get('callbacks')
```

File: tests/checkpointing/test_legacy_checkpoints.py [hunk 6]
```diff
@@ -86,3 +94,34 @@ def test_resume_legacy_checkpoints(tmpdir, pl_version: str):
...
+        default_root_dir=tmpdir,
+        max_steps=1,
+        callbacks=[callback],
+    )
+    trainer.fit(model)
+    # simulate old format where type(callback) was the key
+    new_checkpoint = torch.load(Path(tmpdir, "last.ckpt"))
+    old_checkpiont = deepcopy(new_checkpoint)
+    old_checkpiont["callbacks"] = {type(callback): new_checkpoint["callbacks"]["ModelCheckpoint"]}
+    torch.save(old_checkpiont, Path(tmpdir, "old.ckpt"))
```
Comment (carmocca) [inline tests/checkpointing/test_legacy_checkpoints.py:119]:
```suggestion
    old_checkpoint = deepcopy(new_checkpoint)
    old_checkpoint["callbacks"] = {type(callback): new_checkpoint["callbacks"]["ModelCheckpoint"]}
    torch.save(old_checkpoint, Path(tmpdir, "old.ckpt"))
```

Comment (carmocca) [inline tests/checkpointing/test_legacy_checkpoints.py:119]: Why don't you patch your ModelCheckpoint instance instead?

`callback.state_identifier = callback._legacy_state_identifier`

so you don't have to load and modify the checkpoint manually

**Review by carmocca — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

File: tests/checkpointing/test_legacy_checkpoints.py (hunk 6 above)
Comment (awaelchli) [reply tests/checkpointing/test_legacy_checkpoints.py]: eehm yeah, this test wasn't good. improved now

File: pytorch_lightning/callbacks/base.py [hunk 7]
```diff
@@ -33,6 +33,14 @@ class Callback(abc.ABC):
...
     """
 
+    @property
+    def state_id(self) -> str:
+        return self.__class__.__qualname__
+
+    @property
+    def _legacy_state_id(self) -> Type:
+        return type(self)
+
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/base.py:43]: nit: add docstrings to describe these? otherwise the historical context for the legacy state id will quickly be lost

how long should we support the legacy state id? as long as possible? for X number of PyTorch Lightning versions? should we expect users to use checkpoint upgrade utilities to keep these compliant?

Comment (awaelchli) [reply pytorch_lightning/callbacks/base.py]: added docs

I would prefer as long as possible since we can upgrade the checkpoint dict automatically when it gets loaded. See #8558. Do you have any suggestions of how long?

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/callbacks/base.py [hunk 8]
```diff
@@ -33,6 +33,21 @@ class Callback(abc.ABC):
...
 
+    @property
+    def state_id(self) -> str:
+        """
+        Identifier for the state of the callback. Used to store and retrieve a callback's state from the
+        checkpoint dictionary by ``checkpoint["callbacks"][state_id]``. Implementations of a callback need to
+        provide a unique state id if 1) the callback has state and 2) it is desired to maintain the state of
+        multiple instances of that callback.
+        """
+        return self.__class__.__qualname__
```
Comment (carmocca) [inline pytorch_lightning/callbacks/base.py:44]: This still has the issue that if the class is moved to a different package, the qualname will not match, correct?

File: pytorch_lightning/callbacks/base.py [hunk 9]
```diff
@@ -33,6 +33,21 @@ class Callback(abc.ABC):
...
+
+    @property
+    def _legacy_state_id(self) -> Type:
```
Comment (carmocca) [inline pytorch_lightning/callbacks/base.py:47]:
```suggestion
    def _legacy_state_id(self) -> Type["Callback"]:
```

File: pytorch_lightning/trainer/callback_hook.py [hunk 10]
```diff
@@ -303,7 +302,7 @@ def on_load_checkpoint(self, checkpoint):
             )
 
         for callback in self.callbacks:
-            state = callback_states.get(type(callback))
+            state = callback_states.get(callback.state_id, callback_states.get(callback._legacy_state_id))
```
Comment (carmocca) [inline pytorch_lightning/trainer/callback_hook.py:305]: Should `callback_states.get(callback._legacy_state_id)` be changed to `callback_states[callback._legacy_state_id]`

so we don't silently return `None` and instead fail?

We expect the legacy state id to be defined in all cases

File: tests/callbacks/test_early_stopping.py [hunk 11]
```diff
@@ -76,7 +76,8 @@ def test_resume_early_stopping_from_checkpoint(tmpdir):
     checkpoint = torch.load(checkpoint_filepath)
     # the checkpoint saves "epoch + 1"
     early_stop_callback_state = early_stop_callback.saved_states[checkpoint["epoch"] - 1]
-    assert checkpoint["callbacks"][type(early_stop_callback)] == early_stop_callback_state
+    assert 4 == len(early_stop_callback.saved_states)
+    assert checkpoint["callbacks"]["EarlyStoppingTestRestore"] == early_stop_callback_state
```
Comment (carmocca) [inline tests/callbacks/test_early_stopping.py:80]: The previous code did work, correct?

Just asking, not saying it should be reverted

File: tests/checkpointing/test_legacy_checkpoints.py [hunk 12]
```diff
@@ -95,3 +97,33 @@ def test_resume_legacy_checkpoints(tmpdir, pl_version: str):
...
+        return type(self)
+
+    def on_save_checkpoint(self, *args):
+        return {"state": self.state}
+
+    def on_load_checkpoint(self, trainer, pl_module, callback_state):
+        self.state = callback_state["state"]
+
+
+def test_resume_callback_state_saved_by_type(tmpdir):
```
Comment (carmocca) [inline tests/checkpointing/test_legacy_checkpoints.py:117]: I'm not sure this is the correct file for this test.

Testing the legacy state id is not really related to what we call "legacy checkpoints"

File: tests/checkpointing/test_model_checkpoint.py [hunk 13]
```diff
@@ -858,7 +858,7 @@ def test_model_checkpoint_save_last_checkpoint_contents(tmpdir):
     assert ckpt_last_epoch["epoch"] == ckpt_last["epoch"]
     assert ckpt_last_epoch["global_step"] == ckpt_last["global_step"]
 
-    ch_type = type(model_checkpoint)
+    ch_type = "ModelCheckpoint"
     assert ckpt_last["callbacks"][ch_type] == ckpt_last_epoch["callbacks"][ch_type]
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:862]:
```suggestion
    assert ckpt_last["callbacks"]["ModelCheckpoint"] == ckpt_last_epoch["callbacks"]["ModelCheckpoint"]
```

File: tests/trainer/logging_/test_logger_connector.py [hunk 14]
```diff
@@ -29,7 +30,7 @@
 
 
 def test_fx_validator(tmpdir):
-    funcs_name = sorted(f for f in dir(Callback) if not f.startswith("_"))
+    funcs_name = sorted(m for m, _ in getmembers(Callback, predicate=isfunction) if not m.startswith("_"))
```
Comment (carmocca) [inline tests/trainer/logging_/test_logger_connector.py:33]: Could replace this with

```suggestion
from tests.models.test_hooks import get_members

    funcs_name = sorted(get_members(Callback))
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/callbacks/base.py (hunk 7 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/base.py]: > as long as possible?

I think so, breaking old checkpoints is not cool. Especially if we value reproducibility.

File: pytorch_lightning/callbacks/base.py (hunk 8 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/base.py:44]: if the class is renamed then yes, the qual name will change. if it get's moved it does not matter. 

It will depend who is the author of the callback. If it is us, the PL team, then we can write a migration within Lightning to provide BC, see #8558. If the author is the user, they will have to take care themselves.

Comment (awaelchli) [reply pytorch_lightning/callbacks/base.py:44]: And this here is just the default implementation. See here for how certain callbacks could implement the state_id: https://github.com/PyTorchLightning/pytorch-lightning/pull/7187/files

File: tests/callbacks/test_early_stopping.py (hunk 11 above)
Comment (awaelchli) [reply tests/callbacks/test_early_stopping.py:80]: Yes I checked, on master the length is also 4. I don't remember why I added the length check.

File: tests/checkpointing/test_legacy_checkpoints.py (hunk 12 above)
Comment (awaelchli) [reply tests/checkpointing/test_legacy_checkpoints.py:117]: yes, I partially agree
better to move to tests/callbacks/test_callbacks.py?

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/callback_hook.py (hunk 10 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/callback_hook.py:305]: actually this won't work. We are loading a checkpoint here and we .get() a key but it might be a checkpoint with the new format or the old. So if the checkpoint has legacy keys, `callback_states.get(callback._legacy_state_id)` will return us the right answer, otherwise for new checkpoints above 1.4 we get the element by the new key.
